### PR TITLE
Support Matcha swaps in QuarkBuilder

### PR DESCRIPTION
**Pull request by kevincheng96**
This PR adds support for Matcha swaps in the QuarkBuilder. I also changed the action context for swaps from two separate `Buy` and `Sell` action contexts to one single `Swap` action context.

### CODE REVIEW

**Review by AungThuSoe24 — APPROVED** [review]

**Review by AungThuSoe24 — APPROVED** [review]

File: src/builder/QuarkBuilder.sol [hunk 1]
```diff
@@ -332,6 +332,166 @@ contract QuarkBuilder {
...
+        actions = Actions.truncate(actions, actionIndex);
+        quarkOperations = Actions.truncate(quarkOperations, actionIndex);
+
+        // Validate generated actions for affordability
+        if (payment.isToken) {
+            assertSufficientPaymentTokenBalances(actions, chainAccountsList, swapIntent.chainId);
+        }
+
+        // Construct EIP712 digests
+        EIP712Helper.EIP712Data memory eip712Data;
```
Comment (hayesgm) [inline src/builder/QuarkBuilder.sol:469]: We can share this code at some point, right? In fact, most of the end of this function could (technically, not sure about Solidity-wise) just be a common function call, right?

File: src/builder/QuarkBuilder.sol [hunk 2]
```diff
@@ -332,6 +332,166 @@ contract QuarkBuilder {
...
+        if (quarkOperations.length == 1) {
+            eip712Data = EIP712Helper.EIP712Data({
+                digest: EIP712Helper.getDigestForQuarkOperation(
+                    quarkOperations[0], actions[0].quarkAccount, actions[0].chainId
+                    ),
+                domainSeparator: EIP712Helper.getDomainSeparator(actions[0].quarkAccount, actions[0].chainId),
+                hashStruct: EIP712Helper.getHashStructForQuarkOperation(quarkOperations[0])
+            });
+        } else if (quarkOperations.length > 1) {
```
Comment (hayesgm) [inline src/builder/QuarkBuilder.sol:478]: What about the case, FWIW, where `quarkOperations.length == 0`?

**Review by hayesgm — APPROVED** [review]
dope dope dope

File: src/builder/QuarkBuilder.sol (hunk 2 above)
Comment (kevincheng96) [reply src/builder/QuarkBuilder.sol:478]: In that case, EIP712Data will be an empty struct.

File: src/builder/QuarkBuilder.sol (hunk 1 above)
Comment (kevincheng96) [reply src/builder/QuarkBuilder.sol:469]: Yeah, this logic is being used in multiple places now. We'll want to clean it up a bit in a separate PR.

**Pull request merged**